hdc/algo/accessors.py

```diff
@@ -540,22 +537,37 @@ def spi(
             )
 
         else:
-            groups = np.array(groups) if not isinstance(groups, np.ndarray) else groups
-            num_groups = np.unique(groups).size
 
-            if not groups.dtype.name == "int16":
-                warn("Casting groups to int16!")
-                groups = groups.astype("int16")
+            groups, keys = to_linspace(np.array(groups, dtype="str"))
+
+            if len(groups) != len(self._obj.time):
+                raise ValueError("Need array of groups same length as time dimension!")
+
+            groups = groups.astype("int16")
+            num_groups = len(keys)
+
+            cal_indices = get_calibration_indices(
+                tix, calibration_start, calibration_stop, groups, num_groups
+            )
```

Comment:
the idea here is that if  `groups` is supplied, the `get_calibration_indices` returns a 2-d array, which `len(groups)` rows and 2 columns (`start`, `stop`). This can be then passed on to numba efficiently.

Reply:
In array slicing in Python `stop` typically refers to one past the last element `[start, stop)`, and it's usually an index. Here it seems that `[start, stop]` is used, and also these are not just indices but dates? I think this adds to confusion. Maybe `first,last` or `in,out`, or even `t0, t1` is a better name here, other option is to use a tuple `calibration_{range,period}`?

Reply:
`calibration_{start,stop}` are dates - the function `get_calibration_indices` returns either a tuple or array with containing indices which comply with the python slicing syntax. 

we could rename `calibration_{start,stop}` to make it clearer, and `get_calibration_indices` to `get_calibration_range`?